src/helpers/proposed_network.rs

```diff
@@ -0,0 +1,198 @@
+    #[must_use]
+    pub fn new(query_id: QueryId, callback: oneshot::Sender<()>) -> Self {
+        PrepareQueryData { query_id, callback }
+    }
+
+    pub fn respond(self) -> Result<(), NetworkEventError> {
+        let query_id = self.query_id;
+        self.callback
+            .send(())
+            .map_err(|_| NetworkEventError::CallbackFailed {
+                event_name: "PrepareQuery",
+                query_id,
+            })
+    }
+}
+
+pub struct StartMulData {
+    pub query_id: QueryId,
+    pub endpoints_positions: [Uri; 3],
+    pub field_type: String,
+    pub data_stream: ByteArrStream,
+    callback: oneshot::Sender<()>,
+}
+
+impl StartMulData {
+    pub fn new(
+        query_id: QueryId,
+        endpoints_positions: [Uri; 3],
+        field_type: String,
+        data_stream: ByteArrStream,
+        callback: oneshot::Sender<()>,
+    ) -> Self {
+        StartMulData {
+            query_id,
+            endpoints_positions,
+            field_type,
+            data_stream,
+            callback,
+        }
+    }
+
+    pub fn respond(self) -> Result<(), NetworkEventError> {
+        let query_id = self.query_id;
+        self.callback
+            .send(())
+            .map_err(|_| NetworkEventError::CallbackFailed {
+                event_name: "StartMul",
+                query_id,
+            })
+    }
+}
+
+pub struct MulData {
+    pub query_id: QueryId,
+    pub endpoints_positions: [Uri; 3],
+    pub endpoints_to_roles_and_stream: HashMap<Uri, (Role, ByteArrStream)>,
+}
+
+impl MulData {
+    pub fn new(
+        query_id: QueryId,
+        endpoints_positions: [Uri; 3],
+        endpoints_to_roles_and_stream: HashMap<Uri, (Role, ByteArrStream)>,
+    ) -> Self {
+        Self {
+            query_id,
+            endpoints_positions,
+            endpoints_to_roles_and_stream,
+        }
+    }
+}
+
+pub struct RingEventData {
+    pub query_id: QueryId,
+    pub roles_to_endpoints: HashMap<Role, Uri>,
+    pub ring_event: RingEvent,
+}
+
+impl RingEventData {
+    pub fn new(
+        query_id: QueryId,
+        roles_to_endpoints: HashMap<Role, Uri>,
+        ring_event: RingEvent,
+    ) -> Self {
+        Self {
+            query_id,
+            roles_to_endpoints,
+            ring_event,
+        }
+    }
+}
+
+pub enum NetworkCommand {
+    // Commands sent to handle query creation and initialization
+    CreateQuery(CreateQueryData),
+    PrepareQuery(PrepareQueryData),
+    StartMul(StartMulData),
+    Mul(MulData),
+
+    // Commands sent within the context of a `Ring`, to be used internally
+    RingEvent(RingEventData),
+}
+
+pub trait Network<R: Ring> {
+    type EventStream: Stream<Item = NetworkCommand>;
+    type Sink: futures::Sink<NetworkCommand, Error = NetworkEventError>;
+
+    /// To be called by the entity which will handle events being emitted by `Network`.
+    /// # Panics
+    /// May panic if called more than once
+    fn register(&self) -> Self::EventStream;
```

Comment:
would it allow QPL to sign up for query events only?

Reply:
would something like this work?

```rust
/// returns a handle that allows receiving the specified event type. Note that there is only one handle per event type 
/// so it may not be available when requested. Alternatively this method could panic
fn subscribe(&self, subscription: EventSubscription) -> Option<Self::EventStream> { ... }


enum EventSubscription {
     /// Prepare, NewQuery, etc.
     QueryEvent(EventType),
     /// what is currently used by Infra is here. Events are partitioned by query id, so we must be able to call this multiple times, each time using a unique query id
     StepDataEvent(QueryId)
}
```

Reply:
hm, i see. the only (still) weird part about this is `StepDataEvent` would never be used externally. ie the only entity that would subscribe to this event type is the network itself.

Reply:
true, but that would be another layer on top of transport, so it would make sense to use the frontdoor for it too